docs/reference/src/language/syntax/functions.md

```diff
@@ -14,27 +32,122 @@ Functions can be annotated with the `pure` keyword.
+## Calling a function
+
+A function can be called without a target (like a function call in other languages) or with
+a target (like a method call in other languages):
+
+```slint,no-preview
+export component Example {
+    property <string> my-property: my-function(); // Called without a target
+    property <int> my-other-property: my_button.my-other-function(); // Called with a named target
+
+    pure function my-function() -> string {
+        return "result";
+    }
+
+    Text {
+        text: root.my-function(); // Called with a pre-defined name
+    }
+
+    my_button := Button {
+        pure function my-other-function() -> int {
+            return 42;
+        }
+    }
+}
+```
+
+The two types of calls function mostly the same. The only differences are:
+
+- Functions can be called without a target only from the element that defines them and its children.
+That means a function defined at the root of a component can be called from anywhere within that component
+while a function defined in an element can be called from that element and any child elements of it.
+- When a function is overridden, calling it with a target allows the caller to choose the exact version
+it wants to call (see Function Overriding below).
+
+By default, functions are private and cannot be accessed from other components.
+
 However, their accessibility can be modified using the `public` or `protected` keywords.
 
-- A function annotated with `public` can be accessed by any component.
+- A root-level function annotated with `public` can be accessed by any component.
+
+To access such a function from a different component, you always need a target, which in practice
+means the calling component must declare the called component as one of its child elements.
+
+```slint,no-preview
+export component HasFunction {
+    public pure function double(x: int) -> int {
+        return x * 2;
+    }
+}
+
+export component CallsFunction {
+    property <int> test: my-friend.double(1);
+
+    my-friend: HasFunction {
+    }
+}
+```
+
+If a function is declared in a child element, even if marked public, it is not possible to call it
+from another component, as the child elements themselves are not public and a valid target for the
+function does not exist:
+
+```slint,no-preview
+export component HasFunction {
+    t := Text {
+        public pure function double(x: int) -> int {
+            return x * 2;
+        }
+    }
+}
+
+export component CallsFunction {
+    // Compiler error!
+    // property <int> test: my-friend.t.double(1);
+
+    my-friend: HasFunction {
+    }
+}
+```
+
 - A function annotated with `protected` can only be accessed by components that directly inherit from it.
 
-## Example
+Functions, even marked public, cannot be exported and cannot be called from backend code (in Rust, C++,
+JS, etc.). Declare a [callback](callbacks.md) which can call the function.
+
+## Function overriding
```

Comment:
I'm not sure this is technically overriding. It is just name scoping.
Same is true for callback or properties with the same name.

Reply:
It's not really "scoping" either, I think, because the scope of the function is still the whole component. But I reorganized things a bit and describe this as name resolution rules, and instead of overriding I talk about "shadowing" -- this is similar to variable shadowing in Rust (https://doc.rust-lang.org/rust-by-example/variable_bindings/scope.html).

I think it's clearer this way, please take a look!